lib/controller.rb

```diff
@@ -1,21 +1,47 @@
 require_relative 'game_factory.rb'
 
 class Controller
-  def initialize(display, game_factory)
+  def initialize(display, game_factory, game_saver, game_validator, filename)
     @display = display
     @game_factory = game_factory
+    @game_saver = game_saver
+    @game_validator = game_validator
+    @filename = filename
   end
 
   def main_game
     game_setup
-    play_move until @game.over?
+    until @game.over?
+      play_move
+      save_game?
+      exit_game?
+    end
     end_of_game
   end
 
   private
 
-  def game_setup(squares = [1, 2, 3, 4, 5, 6, 7, 8, 9])
-    @game = @game_factory.create_game(squares)
+  def save_game?
+    if @game.save_game
+      @game.reset_save_game
+      game_name = @game_validator.new_game_name(@filename)
+      @game_saver.save(@filename, game_name, @game)
+      @display.save_game_confirmation
+    end
+  end
+
+  def exit_game?
+    goodbye if @game.exit_game
+  end
+
+  def goodbye
+    @display.show_exit_message
+    exit(0)
```

Comment:
Could this use the game loop to exit? It adds complexity to have a game loop that can be exited directly in a method call within the game loop. At this stage it is manageable, but imagine if there were more ways to exit the game loop - how easy would it be to follow all possible combinations?

Reply:
Good point. I'll look into this, thanks :)